File: RoyalVNCObjCDemo/Source/Connection/ConnectionWindowController.m
#import "ConnectionWindowController.h"

#import "ConnectionViewController.h"
#import "ConnectionViewControllerDelegate.h"

@interface ConnectionWindowController () <NSWindowDelegate, ConnectionViewControllerDelegate>

@property (readwrite, strong) VNCConnectionSettings* settings;
@property (strong) ConnectionViewController* connectionViewController;

@property BOOL didLoad;

@end

@implementation ConnectionWindowController

- (NSNibName)windowNibName {
	return @"ConnectionWindow";
}

- (instancetype)initWithSettings:(VNCConnectionSettings *)settings {
	self = [super initWithWindow:nil];
	
	if (self) {
		self.settings = settings;
		
		self.connectionViewController = [[ConnectionViewController alloc] initWithSettings:settings];
		self.connectionViewController.delegate = self;
	}
	
	return self;
}

- (void)dealloc {
	self.delegate = nil;
}

- (void)windowDidLoad {
	[super windowDidLoad];
	
	if (self.didLoad) {
		return;
	}
	
	self.didLoad = YES;
	
	if (!self.window ||
		!self.window.contentView) {
		return;
	}
	
	self.window.title = [NSString stringWithFormat:@"Connection to %@", self.settings.hostname];
	
	NSView* connectionView = self.connectionViewController.view;
	connectionView.frame = self.window.contentView.bounds;
	connectionView.autoresizingMask = NSViewMinXMargin | NSViewMaxXMargin | NSViewMinYMargin | NSViewMaxYMargin | NSViewWidthSizable | NSViewHeightSizable;

	[self.window.contentView addSubview:connectionView];
}

- (void)connect {
	[self.connectionViewController connect];
}

- (void)disconnect {
	[self.connectionViewController disconnect];
}

- (BOOL)windowShouldClose:(NSWindow*)sender {
	[self disconnect];
	
	return NO;
}

- (VNCConnection*)connection {
	return self.connectionViewController.connection;
}

- (void)connectionViewControllerDidDisconnect:(ConnectionViewController*)connectionViewController {
	self.connectionViewController.delegate = nil;

	[self close];

	[self.delegate connectionWindowControllerDidClose:self];
}

@end

File: RoyalVNCDemo/Source/Connection/ConnectionWindowController.swift
import Foundation
import AppKit

import RoyalVNCKit

class ConnectionWindowController: NSWindowController {
	override var windowNibName: NSNib.Name? { "ConnectionWindow" }
	
	let settings: VNCConnection.Settings
	weak var delegate: ConnectionWindowControllerDelegate?
	
	var connection: VNCConnection { connectionViewController.connection }
	
	private var didLoad = false
	
	private let connectionViewController: ConnectionViewController
	
	init(settings: VNCConnection.Settings) {
		self.settings = settings
		connectionViewController = .init(settings: settings)
		
		super.init(window: nil)
		
		connectionViewController.delegate = self
	}

	@available(*, unavailable)
	required init?(coder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}
	
	deinit {
		connectionViewController.delegate = nil
	}
	
	override func windowDidLoad() {
		guard !didLoad else { return }
		didLoad = true
		
		guard let window = window,
			  let contentView = window.contentView else {
			return
		}
		
		window.title = "Connection to \(settings.hostname)"
		
		let connectionView = connectionViewController.view
		connectionView.frame = contentView.bounds
		connectionView.autoresizingMask = [ .minXMargin, .maxXMargin, .minYMargin, .maxYMargin, .width, .height ]
		
		contentView.addSubview(connectionView)
	}
	
	func connect() {
		connectionViewController.connect()
	}
	
	func disconnect() {
		connectionViewController.disconnect()
	}
}

extension ConnectionWindowController: NSWindowDelegate {
	func windowShouldClose(_ sender: NSWindow) -> Bool {
		disconnect()
		
		return false
	}
}

extension ConnectionWindowController: ConnectionViewControllerDelegate {
	func connectionViewControllerDidDisconnect(_ connectionViewController: ConnectionViewController) {
		connectionViewController.delegate = nil
		
		close()
		
		delegate?.connectionWindowControllerDidClose(self)
	}
}
